### PR TITLE
[crypto-js] Update crypto-js v4.0.0

### PR DESCRIPTION
**Pull request by SevenOutman**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://cryptojs.gitbook.io/docs/
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [x] Include [tests for your changes](https://github.com/DefinitelyTyped/DefinitelyTyped#testing)
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

### CODE REVIEW

File: types/crypto-js/index.d.ts [hunk 1]
```diff
@@ -1,171 +1,1731 @@
...
+                 *
+                 * @throws Error If neither the formatter nor the default formatter is set.
+                 *
+                 * @example
+                 *
+                 *     var string = cipherParams + '';
+                 *     var string = cipherParams.toString();
+                 *     var string = cipherParams.toString(CryptoJS.format.OpenSSL);
+                 */
+                toString(formatter?: Format): string;
```
Comment (mymindstorm) [inline types/crypto-js/index.d.ts:758]: Please double check if this can accept Encoder.

File: types/crypto-js/index.d.ts [hunk 2]
```diff
@@ -1,171 +1,1731 @@
...
-    interface Decryptor {
-        process(messagePart: string): string;
-        finalize(): string;
-    }
-    interface LibWordArray {
-        sigBytes: number,
-        words: number[],
-        toString(encoder?: Encoder): string;
-    }
-    export interface WordArray {
```
Comment (mymindstorm) [inline types/crypto-js/index.d.ts:50]: This isn't exported at the top level anymore, and might break things if removed.

**Review by mymindstorm — CHANGES_REQUESTED** [review]
Please remove the v3 folder

File: types/crypto-js/index.d.ts (hunk 1 above)
Comment (SevenOutman) [reply types/crypto-js/index.d.ts:758]: Not really. `Encoder` can only stringify `WordArray`, not `CipherParams`.

File: types/crypto-js/index.d.ts (hunk 2 above)
Comment (SevenOutman) [reply types/crypto-js/index.d.ts:50]: I don't quite understand. Do you mean `WordArray` should or shouldn't be exported at the top level?

Comment (mymindstorm) [reply types/crypto-js/index.d.ts:50]: Yes, previous versions of these types did that. I'll approve when you remove the v3 folder.

**Review by mymindstorm — APPROVED** [review]

**Comment by SevenOutman** [comment]
Ready to merge

**Comment by tomaszzmudzinski** [comment]
Hi guys, I am wondering why this not work after @types/crypto-js@4.0.0 upgrade:

return CryptoJS.AES.decrypt(textToDecrypt, SECRET_KEY.trim()).toString(CryptoJS.enc.Utf8);

image

Could you tell me how to do it right now (after the library update)

**Comment by SevenOutman** [comment]
@tomaszzmudzinski `AES.decrypt()` returns a `CipherParams` which does not support `.toString(Encoder)` in `crypto-js@4`. You can call `.toString(Encoder)` on its `ciphertext` which is a `WordArray`.
```js
CryptoJS.AES.decrypt(textToDecrypt, SECRET_KEY.trim()).ciphertext.toString(CryptoJS.enc.Utf8);
```

**Comment by hieplpvip** [comment]
@SevenOutman I got "ciphertext is undefined". There is something wrong with @types/crypto-js 4.0.0. @types/crypto-js 3.1.47 works fine with crypto-js 4.0.0

**Comment by SevenOutman** [comment]
@hieplpvip Looks like a mistake [in this PR](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/47687/files#diff-f7fa2d4559992e1ca076949f98f71ca3R274), where return type of `decrypt()` was [supposed to be `WordArray`](https://github.com/brix/crypto-js/blob/4.0.0/cipher-core.js#L720) but incorrectly typed `CipherParams`. I'm committing a fix to this.

**Pull request merged**